Homework7/task2/tree.c

```diff
@@ -0,0 +1,219 @@
+	struct Node* rightChild;
+} Node;
+
+typedef struct Tree {
+	Node* root;
+} Tree;
+
+void addToTreeRecursion(FILE* file, Node* node, int* errorCode) {
+	getc(file);
+	char* operator = calloc(2, sizeof(char));
+	if (operator == NULL) {
+		*errorCode = -1;
+		return;
+	}
+	operator[0] = getc(file);
+	node->value = operator;
+	getc(file);
+	char* nextSymbol = calloc(2, sizeof(char));
+	nextSymbol[0] = (char)getc(file);
+	if (nextSymbol[0] == '(') {
+		ungetc(nextSymbol[0], file);
+		Node* newLeftNode = calloc(1, sizeof(Node));
+		if (newLeftNode == NULL) {
+			free(operator);
+			free(nextSymbol);
+			*errorCode = -1;
+			return;
+		}
+		node->leftChild = newLeftNode;
+		addToTreeRecursion(file, newLeftNode, errorCode);
+		if (*errorCode == -1) {
+			free(newLeftNode);
+			free(operator);
+			free(nextSymbol);
+			return;
+		}
+	} else {
+		ungetc(nextSymbol[0], file);
+		node->leftChild = calloc(1, sizeof(Node));
+		if (node->leftChild == NULL) {
+			*errorCode = -1;
+			return;
+		}
+		node->leftChild->value = calloc(MAX_NUMBER_SIZE + 1, sizeof(char));
+		if (node->leftChild->value == NULL) {
+			free(node->leftChild);
+			*errorCode = -1;
+			return;
+		}
+		fscanf(file, "%s", node->leftChild->value);
+	}
+	getc(file);
+	nextSymbol[0] = (char)getc(file);
+	if (nextSymbol[0] == '(') {
+		ungetc(nextSymbol[0], file);
+		Node* newRightNode = calloc(1, sizeof(Node));
+		if (newRightNode == NULL) {
+			free(operator);
+			free(nextSymbol);
+			*errorCode = -1;
+			return;
+		}
+		node->rightChild = newRightNode;
+		addToTreeRecursion(file, newRightNode, errorCode);
+		if (*errorCode == -1) {
+			free(newRightNode);
+			free(operator);
+			free(nextSymbol);
+			return;
+		}
+	} else {
+		ungetc(nextSymbol[0], file);
+		node->rightChild = calloc(1, sizeof(Node));
+		if (node->rightChild == NULL) {
+			free(operator);
+			free(nextSymbol);
+			*errorCode = -1;
+			return;
+		}
+		node->rightChild->value = calloc(MAX_NUMBER_SIZE + 1, sizeof(char));
+		if (node->rightChild->value == NULL) {
+			free(node->rightChild);
+			free(operator);
+			free(nextSymbol);
+			*errorCode = -1;
+			return;
+		}
+		fscanf(file, "%[^)]", node->rightChild->value);
+		getc(file);
+	}
+}
+
+int readFileToTree(const char* fileName, Tree* tree) {
+	FILE* file = fopen(fileName, "r");
+	if (file == NULL) {
+		return -1;
+	}
+	tree->root = malloc(sizeof(Node));
+	if (tree->root == NULL) {
+		fclose(file);
+		return -1;
+	}
+	int errorCode = 0;
+	addToTreeRecursion(file, tree->root, &errorCode);
+	if (errorCode == -1) {
+		fclose(file);
+		deleteTree(&tree);
+		return -1;
+	}
+	fclose(file);
+	return 0;
+}
+
+Tree* createParseTree(void) {
+	Tree* tree = calloc(1, sizeof(Tree));
+	return tree;
+}
+
+void deleteTreeRecursive(Node* node) {
+	if (node == NULL) {
+		return;
+	}
+	deleteTreeRecursive(node->leftChild);
+	deleteTreeRecursive(node->rightChild);
+	free(node->value);
+	free(node);
+	node = NULL;
+}
+
+void deleteTree(Tree** tree) {
+	if ((*tree) == NULL) {
```

Comment:
```suggestion
	if (*tree == NULL) {
```
И ниже. Я понимаю, что так копипастить удобнее, но так не пишут